File: maximum-depth-of-binary-tree/maximum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: TreeNode) -> int:
        if not root:
            return 0
        return self.dfs(root, 0)
    
    def dfs(self, root: TreeNode, depth: int) -> int:
        if not root:
            return depth
        depth += 1
        return max(self.dfs(root.left, depth), self.dfs(root.right, depth))

File: maximum-depth-of-binary-tree/maximum-depth-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution 
{
    public int maxDepth(TreeNode root) 
    {
        int depth = 0;
        if (root == null)
            return depth;
        
        Queue<TreeNode> queue = new LinkedList<TreeNode>();
        queue.add(root);
        int currentLevel = 1;
        int nextLevel = 0;
        while (currentLevel > 0)
        {
            TreeNode node = queue.remove();
            currentLevel--;
            if (node.left != null)
            {
                queue.add(node.left);
                nextLevel++;
            }
            if (node.right != null)
            {
                queue.add(node.right);
                nextLevel++;
            }
            
            if (currentLevel == 0)
            {
                currentLevel = nextLevel;
                nextLevel = 0;
                depth++;
            }
        }
        return depth;
    }
}